### ISSUE TITLE
Cannot connect to my Cosmos endpoint from MapReduce application

### ISSUE THREAD

**Issue by benhsu** [Azure/azure-sdk-for-java#27206]
**Query/Question**
I am using the SDK in my Hadoop MapReduce application to post the results to a CosmosDB, and I am getting an error.

In the setup() function of my Reducer, I have this code:
```
        CosmosClient cosmos = new CosmosClientBuilder()
                .key("myKey")
                .endpoint("https://benhsu-cosmos-tryout.documents.azure.com:443/") 
            .buildClient();
```

and when I run it, I get this stack trace:
```
Error: java.lang.RuntimeException: Client initialization failed. Check if the endpoint is reachable and if your auth token is valid. More info: https://aka.ms/cosmosdb-tsg-service-unavailable-java
	at com.azure.cosmos.implementation.RxDocumentClientImpl.initializeGatewayConfigurationReader(RxDocumentClientImpl.java:401)
	at com.azure.cosmos.implementation.RxDocumentClientImpl.init(RxDocumentClientImpl.java:436)
	at com.azure.cosmos.implementation.AsyncDocumentClient$Builder.build(AsyncDocumentClient.java:245)
	at com.azure.cosmos.CosmosAsyncClient.<init>(CosmosAsyncClient.java:127)
	at com.azure.cosmos.CosmosClientBuilder.buildAsyncClient(CosmosClientBuilder.java:773)
	at com.azure.cosmos.CosmosClient.<init>(CosmosClient.java:32)
	at com.azure.cosmos.CosmosClientBuilder.buildClient(CosmosClientBuilder.java:785)
```

***What I have tried***

1. As per https://aka.ms/cosmosdb-tsg-service-unavailable-java , I logged into the worker node (`wn0-benhsu.4iaazpxvmzpexkmpw4balrlqpa.bx.internal.cloudapp.net`) and tried to `telnet benhsu-cosmos-tryout.documents.azure.com 443` and I was able to get a prompt back. Conclusion: no firewall issue
2. I took the "Azure Cosmos Java Getting Started" project (https://github.com/Azure-Samples/azure-cosmos-java-getting-started.git) , cloned it to the above worker node, and ran it with my hostname and key, and it worked. Conclusion: key is fine
3. in my Getting Started project from (2), I tried changing the key to a wrong key and got an error message, conclusion: its probably not my key that is wrong

So its not the key, not the hostname, and not the firewall, can you suggest some steps to figure out what else it could be? Do you have any sample code for using Cosmos SDK from MapReduce?

***Why is this not a Bug or a feature Request?***
I believe the code should work, but I have tried the troubleshooting/debugging steps provided and have not gotten anywhere.

**Setup (please complete the following information if applicable):**
 - OS: Ubuntu 18.04 (what Azure HDI Worker Nodes are running)
 - IDE: IntelliJ
 - Library/Libraries: Hadoop 2.6.0, Cosmos SDK version 4.26.0
 
 **Information Checklist**

**Comment by benhsu** [comment]
forgot to mention: this seems related to https://github.com/Azure/azure-sdk-for-java/issues/17866 , but the reporter there was connecting to an emulator

**Comment by kushagraThapar** [comment]
@benhsu  - that is older sample, approach to connecting mapreduce would be similar, rather using new V4 Java SDK. 
It could also be related to DIRECT mode ports. I would recommend running the cosmosClient in GATEWAY mode, and see if it works fine. 
If it does, then you will need to open ports from your worker node to be able to connect to Cosmos db. See this for more details - https://docs.microsoft.com/en-us/azure/cosmos-db/sql/sql-sdk-connection-modes#service-port-ranges

**Comment by benhsu** [comment]
@kushagraThapar 
Thank you, I will try this
Is this how I use Gateway mode?
```
        CosmosClient cosmos = new CosmosClientBuilder()
            .gatewayMode()
            .endpoint(cosmosEndpoint)
            .key(cosmosKey)
            .buildClient();
```

Also I tried some ports in the 10000 to 20000 range per the service port article
```
sshuser@wn0-benhsu:~$ telnet benhsu-cosmos-tryout.documents.azure.com 10500
Trying 40.78.226.50... 
^C ### FAIL
sshuser@wn0-benhsu:~$ telnet benhsu-cosmos-tryout.documents.azure.com 15000
Trying 40.78.226.50...
Connected to cdb-ms-prod-eastus1-fd78.cloudapp.net.
Escape character is '^]'. ### SUCCESS
```

**Comment by kushagraThapar** [comment]
@benhsu - yes, that's the correct way of using gateway mode. I am surprised if the standalone application works, mapreduce should also work in your worker node. 
Worth trying out the sample application with both direct and gateway mode, and then mapreduce application in the same way.

**Comment by benhsu** [comment]
Hello. I looked more closely at my logs and found some classpath problems with Netty. 
I was also able to connect from my worker node to Cosmos by modifying this code: https://docs.microsoft.com/en-us/azure/hdinsight/hadoop/apache-hadoop-develop-deploy-java-mapreduce-linux 

I will try to use what I learned from  that demo code to get my actual job running. Will reach out if I get stuck resolving classpath issues
